toolchain/check/testdata/impl/assoc_const_self.carbon

```diff
@@ -0,0 +1,482 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// AUTOUPDATE
+// TIP: To test this file alone, run:
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/check/testdata/impl/assoc_const_self.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/impl/assoc_const_self.carbon
+
+// --- assoc_const_self.carbon
+
+library "[[@TEST_NAME]]";
+
+interface I {
+  let V:! Self;
+}
+
+impl {} as I where .V = {} {}
+
+impl i32 as I where .V = 0 {}
```

Comment:
No test with this `I where .V = ...` as a function parameter constraint.

```
fn F(T:! I where {} impls ImplicitAs(.Self) and .V = {});
```

We don't have the support for everything in this expression, but we shouldn't crash.

Reply:
Added.